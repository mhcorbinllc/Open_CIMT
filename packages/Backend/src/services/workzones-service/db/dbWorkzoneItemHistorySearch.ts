import {
  DynaMongoDB,
  findLikeSearchEngine,
} from "dyna-mongo-db";
import {
  IWorkzoneHistory,
  IDBWorkzoneHistory,
  convertIDBWorkzoneHistorytoIWorkzoneHistory,
} from "../interfaces/IWorkzoneHistory";
import {getWorkzonesItemsHistoryCollection} from "./getWorkzonesCollections";

export const dbWorkzoneItemHistorySearch = async (
  {
    dmdb,
    companyId,
    workzoneId,
    searchText,
    skip,
    limit,
  }: {
    dmdb: DynaMongoDB;
    companyId: string;
    workzoneId: string;
    searchText?: string;
    skip: number;
    limit: number;
  },
): Promise<IWorkzoneHistory[]> => {
  const collection = await getWorkzonesItemsHistoryCollection(dmdb, companyId);

  const workzoneHistory: IDBWorkzoneHistory[] = await collection
    .find(
      {
        "workzone.workzoneId": workzoneId,
        ...findLikeSearchEngine('dbSearchText', searchText),
      },
      {
        skip,
        limit,
      },
    )
    .toArray();

  return workzoneHistory.map(convertIDBWorkzoneHistorytoIWorkzoneHistory);
};
