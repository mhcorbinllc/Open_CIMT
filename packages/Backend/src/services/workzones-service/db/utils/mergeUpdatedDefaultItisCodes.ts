import {Db} from "mongodb";
import {syncPromises} from "dyna-sync";

import {
  IWorkzonesAppSettings,
} from "../../interfaces/IWorkzonesAppSettings";
import {getDefaultAppItisCodes} from "../../interfaces/getDefaultAppItisCodes";

export const mergeUpdatedDefaultItisCodes = async (
  {
    db,
    collectionName,
  }: {
    db: Db;
    collectionName: string;
  },
): Promise<void> => {
  const collection = db.collection<IWorkzonesAppSettings>(collectionName);
  const appsSettings = await collection.find().toArray();
  const defaultAppItisCodes = getDefaultAppItisCodes();

  await syncPromises(
    ...appsSettings.map(appSettings => {

      const hasSameCodes =
        appSettings.itisCodes.map(v => v.code).sort()
          .join(',') ===
        defaultAppItisCodes.map(v => v.code).sort()
          .join(',');

      if (hasSameCodes) return () => Promise.resolve();

      const mergedItisCodes = defaultAppItisCodes
        .filter((v) => !appSettings.itisCodes.find(t =>
          t.code === v.code
          && t.label.toLowerCase() === v.label.toLowerCase(),
        ))
        .concat(appSettings.itisCodes);

      return async () => {
        await collection.updateOne(
          {deviceId: appSettings.deviceId},
          {$set: {itisCodes: mergedItisCodes}},
        );
      };
    }),
  );
};
