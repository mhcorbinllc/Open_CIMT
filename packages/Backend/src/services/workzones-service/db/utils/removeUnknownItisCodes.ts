import {IDBWorkzone} from "../../interfaces/IDBWorkzone";
import {DynaMongoDB} from "dyna-mongo-db";
import {dbWorkzonesAppSettingsLoad} from "../dbWorkzonesAppSettingsLoad";

export const removeUnknownItisCodes = async (
  {
    dmdb,
    companyId,
    dbWorkzones,
  }: {
    dmdb: DynaMongoDB;
    companyId: string;
    dbWorkzones: IDBWorkzone[];
  },
): Promise<void> => {
  const settings = await dbWorkzonesAppSettingsLoad(dmdb, companyId);
  dbWorkzones
    .forEach(dbWorkzone => {
      dbWorkzone.itisCodes =
        dbWorkzone.itisCodes
          .filter(itisCode => settings.itisCodes.find(ic => itisCode === ic.code));
    });
};
