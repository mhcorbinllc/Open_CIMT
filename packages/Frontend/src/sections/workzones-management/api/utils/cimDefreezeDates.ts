import {IWorkzone} from "mhc-server";

export const cimDefreezeDates = (cim: IWorkzone): IWorkzone => {
  return {
    ...cim,
    start: new Date(cim.start),
    end: new Date(cim.end),
  };
};
