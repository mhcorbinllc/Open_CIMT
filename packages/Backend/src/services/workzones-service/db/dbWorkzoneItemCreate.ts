import {DynaMongoDB} from "dyna-mongo-db";
import {dynaError} from "dyna-error";
import {guid} from "dyna-guid";

import {IWorkzone} from "../interfaces/IWorkzone";
import {
  IDBWorkzone,
  convertIWorkzoneToIDBWorkzone,
  convertIDBWorkzoneToIWorkzone,
} from "../interfaces/IDBWorkzone";

import {getWorkzonesItemsCollection} from "./getWorkzonesCollections";
import {dbWorkzoneItemHistoryAdd} from "./dbWorkzoneItemHistoryAdd";
import {removeUnknownItisCodes} from "./utils/removeUnknownItisCodes";

export const dbWorkzoneItemCreate = async (
  {
    dmdb,
    companyId,
    userId,
    workzone,
    externalImport = false,
  }: {
    dmdb: DynaMongoDB;
    companyId: string;
    userId: string;
    workzone: IWorkzone;
    externalImport?: boolean;
  },
): Promise<IWorkzone> => {
  const collection = await getWorkzonesItemsCollection(dmdb, companyId);
  const dbWorkzone: IDBWorkzone = convertIWorkzoneToIDBWorkzone(workzone);

  if (externalImport) {
    [
      'workzoneId',
      'createdAt',
      'updatedAt',
      'deletedAt',
    ].forEach(fieldName => {
      if (workzone[fieldName] === undefined) {
        throw dynaError({
          message: `dbWorkzoneItemCreate: [${fieldName}] is not defined`,
          status: 400,
          validationErrors: {[fieldName]: 'Not defined'},
        });
      }
    });
  }
  else {
    dbWorkzone.workzoneId = guid();
    const now = Date.now();
    dbWorkzone.createdAt = now;
    dbWorkzone.updatedAt = now;
    dbWorkzone.deletedAt = 0;
  }

  await collection.insertOne(dbWorkzone);
  await dbWorkzoneItemHistoryAdd({
    dmdb,
    companyId,
    userId,
    historyName: 'CIM creation',
    dbWorkzone: dbWorkzone,
  });

  await removeUnknownItisCodes({
    dmdb,
    companyId,
    dbWorkzones: [dbWorkzone],
  });

  return convertIDBWorkzoneToIWorkzone(dbWorkzone);
};
