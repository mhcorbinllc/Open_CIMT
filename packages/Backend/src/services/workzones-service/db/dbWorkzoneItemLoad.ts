import {IWorkzone} from "../interfaces/IWorkzone";
import {convertIDBWorkzoneToIWorkzone} from "../interfaces/IDBWorkzone";
import {DynaMongoDB} from "dyna-mongo-db";
import {getWorkzonesItemsCollection} from "./getWorkzonesCollections";
import {removeUnknownItisCodes} from "./utils/removeUnknownItisCodes";

export const dbWorkzoneItemLoad = async (
  {
    dmdb,
    companyId,
    workzoneId,
  }: {
    dmdb: DynaMongoDB;
    companyId: string;
    workzoneId: string;
  },
): Promise<IWorkzone | null> => {
  const collection = await getWorkzonesItemsCollection(dmdb, companyId);

  const dbWorkzone = await collection.findOne({workzoneId});
  if (!dbWorkzone) return null;

  await removeUnknownItisCodes({
    dmdb,
    companyId,
    dbWorkzones: [dbWorkzone],
  });

  return convertIDBWorkzoneToIWorkzone(dbWorkzone);
};
