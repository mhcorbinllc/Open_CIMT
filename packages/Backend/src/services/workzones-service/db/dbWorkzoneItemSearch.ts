import {DynaMongoDB} from "dyna-mongo-db";

import {IWorkzone} from "../interfaces/IWorkzone";
import {
  IDBWorkzone,
  convertIDBWorkzoneToIWorkzone,
} from "../interfaces/IDBWorkzone";

import {getWorkzonesItemsCollection} from "./getWorkzonesCollections";
import {dbSearchCollection} from "../../../db/DBEntity/utils/dbSearchCollection";
import {removeUnknownItisCodes} from "./utils/removeUnknownItisCodes";

export const dbWorkzoneItemSearch = async (
  {
    dmdb,
    companyId,
    searchText,
    forEver = 'both',
    deleted,
    skip,
    limit,
    sortByFieldName,
    sortDirection,
  }: {
    dmdb: DynaMongoDB;
    companyId: string;
    searchText?: string;
    forEver?: boolean | 'both';
    deleted?: boolean | 'both';
    skip: number;
    limit: number;
    sortByFieldName?: string;
    sortDirection?: 0 | 1 | -1;
  },
): Promise<IWorkzone[]> => {
  const collection = await getWorkzonesItemsCollection(dmdb, companyId);

  const dbWorkzones = await dbSearchCollection<IDBWorkzone>({
    collection,

    searchText,
    deleted,

    find: forEver === 'both'
      ? undefined
      : {forEver},

    skip,
    limit,

    sortByFieldName,
    sortDirection,
  });

  await removeUnknownItisCodes({
    dmdb,
    companyId,
    dbWorkzones,
  });

  return dbWorkzones.map(convertIDBWorkzoneToIWorkzone);
};
