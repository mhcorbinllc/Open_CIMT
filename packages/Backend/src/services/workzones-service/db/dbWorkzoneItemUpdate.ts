import {IWorkzone} from "../interfaces/IWorkzone";
import {convertIWorkzoneToIDBWorkzone} from "../interfaces/IDBWorkzone";
import {DynaMongoDB} from "dyna-mongo-db";
import {getWorkzonesItemsCollection} from "./getWorkzonesCollections";
import {dbWorkzoneItemHistoryAdd} from "./dbWorkzoneItemHistoryAdd";
import {removeUnknownItisCodes} from "./utils/removeUnknownItisCodes";
import {dbWorkzoneItemLoad} from "./dbWorkzoneItemLoad";

export const dbWorkzoneItemUpdate = async (
  {
    dmdb,
    companyId,
    userId,
    workzone,
    externalImport = false,
    newerOnly = false,
    historyName = "CIM update",
  }: {
    dmdb: DynaMongoDB;
    companyId: string;
    userId: string;
    workzone: IWorkzone;
    externalImport?: boolean;
    newerOnly?: boolean;
    historyName?: string;
  },
): Promise<boolean> => {
  const collection = await getWorkzonesItemsCollection(dmdb, companyId);

  const dbWorkzone = convertIWorkzoneToIDBWorkzone(workzone);

  await removeUnknownItisCodes({
    dmdb,
    companyId,
    dbWorkzones: [dbWorkzone],
  });

  if (newerOnly) {
    const currentWorkzone = await dbWorkzoneItemLoad({
      dmdb,
      companyId,
      workzoneId: workzone.workzoneId,
    });
    if (currentWorkzone && currentWorkzone.updatedAt > workzone.updatedAt) {
      return false;
    }
  }

  if (!externalImport) {
    dbWorkzone.updatedAt = Date.now();
  }

  const updateResult = await collection
    .updateOne(
      {workzoneId: dbWorkzone.workzoneId},
      {$set: dbWorkzone},
    );

  await dbWorkzoneItemHistoryAdd({
    dmdb,
    companyId,
    userId,
    historyName,
    dbWorkzone,
  });

  return updateResult.matchedCount === 1;
};
