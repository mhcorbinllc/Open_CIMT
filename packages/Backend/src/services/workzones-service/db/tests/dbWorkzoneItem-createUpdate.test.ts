import {testDMDB} from "../../../../../tests/utils/testDMDB";

import {clearDynamicValuesForSnapshots} from "../../../../utils/clearDynamicValuesForSnapshots";

import {
  IWorkzone,
  getDefaultWorkzone,
} from "../../client";

import {dbWorkzoneItemCreate} from "../dbWorkzoneItemCreate";
import {dbWorkzoneItemSearch} from "../dbWorkzoneItemSearch";
import {dbWorkzoneItemUpdate} from "../dbWorkzoneItemUpdate";
import {dbWorkzoneItemLoad} from "../dbWorkzoneItemLoad";
import {dbWorkzoneItemHistorySearch} from "../dbWorkzoneItemHistorySearch";

describe('CIM Items - Create Update', () => {
  const companyId = 'test-company--test-480246';
  const userId = 'test-user-id-23242';

  const c = (data: any): any => clearDynamicValuesForSnapshots(data, 'workzoneId'.split(','));

  const dmdb = testDMDB;
  const clearDb = async () => {
    try {
      await dmdb.dropCollection([companyId, 'workzones-items'].join('---'));
      await dmdb.dropCollection([companyId, 'workzones-items-history'].join('---'));
    }
    catch (e) {
      console.error('clearDb error:', e);
    }
  };

  beforeAll(async () => {
    await clearDb();
  });
  afterAll(async () => {
    await clearDb();
    await dmdb.disconnect();
  });

  test('Create Update Check CIM item the history', async () => {
    const workzone: IWorkzone = {
      ...getDefaultWorkzone(),
      name: 'Demo CIM Hutton street',
      start: new Date("2017-12-25T21:56:22Z"),
      end: new Date("2017-12-31T22:00:00Z"),
      notes: "Should be ready on new year's eve!",
    };

    const workzoneId = await dbWorkzoneItemCreate({
      dmdb,
      companyId,
      userId,
      workzone,
    });

    expect(!!workzoneId).toBe(true);

    const searchWorkzones = await dbWorkzoneItemSearch({
      dmdb,
      companyId,
      searchText: 'new year eve',
      skip: 0,
      limit: 100,
    });

    expect(searchWorkzones.length).toBe(1);
    expect(c(searchWorkzones)).toMatchSnapshot('Search CIMs');

    if (!searchWorkzones.length) return; // Exit, will fail anyway!

    const updateWorkzone = searchWorkzones[0];
    updateWorkzone.notes += '\nAlmost done';
    updateWorkzone.point.lat = 32.2411;
    updateWorkzone.point.lng = 14.1244;

    await dbWorkzoneItemUpdate({
      dmdb,
      companyId,
      userId,
      workzone: updateWorkzone,
    });

    const loadedWorkzone = await dbWorkzoneItemLoad({
      dmdb,
      companyId,
      workzoneId: updateWorkzone.workzoneId,
    });

    expect(loadedWorkzone).not.toBe(null);
    expect(c(loadedWorkzone)).toMatchSnapshot('Load updated CIM');
    if (loadedWorkzone) expect(loadedWorkzone.workzoneId).toBe(updateWorkzone.workzoneId);

    const history = await dbWorkzoneItemHistorySearch({
      dmdb,
      companyId,
      workzoneId: updateWorkzone.workzoneId,
      skip: 0,
      limit: 100,
    });

    expect(history.length).toBe(2);
    expect(history.map(c)).toMatchSnapshot('History');
  });

});
