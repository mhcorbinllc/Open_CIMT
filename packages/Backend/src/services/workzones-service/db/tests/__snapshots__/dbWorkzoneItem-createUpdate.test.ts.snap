// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CIM Items - Create Update Create Update Check CIM item the history: History 1`] = `
Array [
  Object {
    "date": "---NUMBER>0---",
    "name": "CIM creation",
    "userId": "test-user-id-23242",
    "workzone": Object {
      "active": true,
      "broadcast": Object {
        "selectionByRadius": Object {
          "radiusInMeters": 0,
        },
        "selectionByRegion": Object {
          "region": Object {
            "points": Array [],
          },
        },
        "selectionMode": "REGION",
      },
      "closedLane": "NONE",
      "closedLaneWidthInMeters": 0,
      "closedShoulder": "NONE",
      "closureSelectionMode": "SIMPLE",
      "closureType": "NONE",
      "createdAt": "---NUMBER>0---",
      "deletedAt": "---NUMBER=0---",
      "end": "2017-12-31T22:00:00.000Z",
      "forEver": false,
      "heading": Array [],
      "itisCodes": Array [],
      "kapschId": -1,
      "name": "Demo CIM Hutton street",
      "notes": "Should be ready on new year's eve!",
      "path": Object {
        "points": Array [],
      },
      "point": Object {
        "alt": 0,
        "lat": 0,
        "lng": 0,
      },
      "speedLimitInMiles": 0,
      "start": "2017-12-25T21:56:22.000Z",
      "updatedAt": "---NUMBER>0---",
      "workersPresent": false,
      "workzoneId": "---STRING_NON_EMPTY---",
    },
  },
  Object {
    "date": "---NUMBER>0---",
    "name": "CIM update",
    "userId": "test-user-id-23242",
    "workzone": Object {
      "active": true,
      "broadcast": Object {
        "selectionByRadius": Object {
          "radiusInMeters": 0,
        },
        "selectionByRegion": Object {
          "region": Object {
            "points": Array [],
          },
        },
        "selectionMode": "REGION",
      },
      "closedLane": "NONE",
      "closedLaneWidthInMeters": 0,
      "closedShoulder": "NONE",
      "closureSelectionMode": "SIMPLE",
      "closureType": "NONE",
      "createdAt": "---NUMBER>0---",
      "deletedAt": "---NUMBER=0---",
      "end": "2017-12-31T22:00:00.000Z",
      "forEver": false,
      "heading": Array [],
      "itisCodes": Array [],
      "kapschId": -1,
      "name": "Demo CIM Hutton street",
      "notes": "Should be ready on new year's eve!
Almost done",
      "path": Object {
        "points": Array [],
      },
      "point": Object {
        "alt": 0,
        "lat": 32.2411,
        "lng": 14.1244,
      },
      "speedLimitInMiles": 0,
      "start": "2017-12-25T21:56:22.000Z",
      "updatedAt": "---NUMBER>0---",
      "workersPresent": false,
      "workzoneId": "---STRING_NON_EMPTY---",
    },
  },
]
`;

exports[`CIM Items - Create Update Create Update Check CIM item the history: Load updated CIM 1`] = `
Object {
  "active": true,
  "broadcast": Object {
    "selectionByRadius": Object {
      "radiusInMeters": 0,
    },
    "selectionByRegion": Object {
      "region": Object {
        "points": Array [],
      },
    },
    "selectionMode": "REGION",
  },
  "closedLane": "NONE",
  "closedLaneWidthInMeters": 0,
  "closedShoulder": "NONE",
  "closureSelectionMode": "SIMPLE",
  "closureType": "NONE",
  "createdAt": "---NUMBER>0---",
  "deletedAt": "---NUMBER=0---",
  "end": "2017-12-31T22:00:00.000Z",
  "forEver": false,
  "heading": Array [],
  "itisCodes": Array [],
  "kapschId": -1,
  "name": "Demo CIM Hutton street",
  "notes": "Should be ready on new year's eve!
Almost done",
  "path": Object {
    "points": Array [],
  },
  "point": Object {
    "alt": 0,
    "lat": 32.2411,
    "lng": 14.1244,
  },
  "speedLimitInMiles": 0,
  "start": "2017-12-25T21:56:22.000Z",
  "updatedAt": "---NUMBER>0---",
  "workersPresent": false,
  "workzoneId": "---STRING_NON_EMPTY---",
}
`;

exports[`CIM Items - Create Update Create Update Check CIM item the history: Search CIMs 1`] = `
Array [
  Object {
    "active": true,
    "broadcast": Object {
      "selectionByRadius": Object {
        "radiusInMeters": 0,
      },
      "selectionByRegion": Object {
        "region": Object {
          "points": Array [],
        },
      },
      "selectionMode": "REGION",
    },
    "closedLane": "NONE",
    "closedLaneWidthInMeters": 0,
    "closedShoulder": "NONE",
    "closureSelectionMode": "SIMPLE",
    "closureType": "NONE",
    "createdAt": "---NUMBER>0---",
    "deletedAt": "---NUMBER=0---",
    "end": "2017-12-31T22:00:00.000Z",
    "forEver": false,
    "heading": Array [],
    "itisCodes": Array [],
    "kapschId": -1,
    "name": "Demo CIM Hutton street",
    "notes": "Should be ready on new year's eve!",
    "path": Object {
      "points": Array [],
    },
    "point": Object {
      "alt": 0,
      "lat": 0,
      "lng": 0,
    },
    "speedLimitInMiles": 0,
    "start": "2017-12-25T21:56:22.000Z",
    "updatedAt": "---NUMBER>0---",
    "workersPresent": false,
    "workzoneId": "---STRING_NON_EMPTY---",
  },
]
`;
